service-loadbalancer/gce/cluster_manager.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	gce "k8s.io/kubernetes/pkg/cloudprovider/providers/gce"
+)
+
+const (
+	defaultPort            = 80
+	defaultHealthCheckPath = "/"
+	defaultPortRange       = "80"
+	defaultPortName        = "default-port"
+	defaultHttpHealthCheck = "k8-default-health-check"
+
+	// A single instance-group is created per cluster manager.
+	// Tagged with the name of the controller.
+	instanceGroupPrefix = "k8-ig"
+
+	// A backend is created per nodePort, tagged with the nodeport.
+	// This allows sharing of backends across loadbalancers.
+	backendPrefix = "k8-be"
+
+	// A single target proxy/urlmap/forwarding rule is created per loadbalancer.
+	// Tagged with the name of the IngressPoint.
+	targetProxyPrefix    = "k8-tp"
+	forwardingRulePrefix = "k8-fw"
+	urlMapPrefix         = "k8-um"
+
+	// The gce api uses the name of a path rule to match a host rule.
+	// In the current implementation,
+	hostRulePrefix = "host"
+
+	// State string required by gce library to list all instances.
+	allInstances = "ALL"
+
+	// Used in the test RunServer method to denote a delete request.
+	deleteType = "del"
+
+	// port 0 is used as a signal for port not found/no such port etc.
+	invalidPort = 0
+)
+
+// ClusterManager manages cluster resource pools.
+type ClusterManager struct {
+	ClusterName            string
+	defaultBackendNodePort int64
+	instancePool           NodePool
+	backendPool            BackendPool
+	l7Pool                 LoadBalancerPool
+	// We currently only employ a single cluster wide health check
+	healthChecker HealthChecker
+}
+
+func (c *ClusterManager) shutdown() error {
+	if err := c.l7Pool.Shutdown(); err != nil {
+		return err
+	}
+	if err := c.backendPool.Shutdown(); err != nil {
+		return err
+	}
+	if err := c.instancePool.Shutdown(); err != nil {
+		return err
+	}
+	if err := c.healthChecker.Delete(defaultHttpHealthCheck); err != nil {
+		return err
+	}
+	// TODO: Delete default health check, or is it actually used by the
+	// cluster instance group?
+	return nil
+}
+
+func defaultInstanceGroupName(clusterName string) string {
```

Comment:
nit: I don't really see the win of these functions.

why not delete these functions and just use

```go
strings.Join([]string{instanceGroupPrefix, clusterName}, "-")
``

for example at the call sites.  I think that's actually clearer to read.